### PR TITLE
Stop medication structured phrase

### PR DESCRIPTION
**Pull request by nicoleng12**
This PR address jira issue 1139 to create a new shortcut for #stop medication. Worked with @gregquinn2001 and @daniellee1025 on note parser updates.

**To test this PR** 
The web app:

Open a new note and type #stop medication. Then enter @active medication and choose one of the medications from the list. View the medication charts in the targeted data panel. A medication that has been stopped will not display the chart and the text updates accordingly. See below where the medication chosen was Lisinopril:
image

This also works when you change the medication type on a medication that already has a medication change. For example updating PACLitaxel from reduced to stopped will update the patient and the visualization.

Note parser:

Note 14 was added for the #stop medication shortcut.

Run the following command:
'yarn parse src/noteparser/samples/note14.txt'
The results of parsing the note will be displayed. You will see a medication change object with the type set to "stop" and a flux medication requested of "ibuprofen 600mg tablet"

**Changes made**
- Added #stop medication shortcut and @active medication shortcut
- Updated shortcuts.json (added new attributes) and note parser to support these new shortcuts (*Note: Added infrastructure to support structured phrases using # followed by @ and support parsing of inserters)
- Updated the note parser to handle inserters and added a patient to be able to look up active medications on the patient when parsing @active medication
- Added a new sample note with #stop medication and @active medication 
- Fixed broken backend note parser tests (updated expected jsons to include missing attributes and removed entryid and shrid prior to making comparison)
- Updated Flux objects to set entry information when none exists and added setters 
- Updated MedicationRangeChartVisualizer to change display when medication change is type "stop"
- Refactored some code in SummaryMetadata to more generically handle medication changes

_Pull requests into Flux require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable (**N/A**) and :white_check_mark:._

## Submitter:

**Running the Application**

- [x] Manually tested in Chrome
- [x] Manually tested in IE

**Documentation**

- [x] This pull request describes why these changes were made
- [x] Recognizes any potential shortcomings/bugs in the description 
- N/A Cheat sheet is updated
- N/A Demo script is updated 
- [x] Documentation on Wiki has been updated 
- N/A Additional technical components and libraries are documented and added to wiki as needed

**NoteParser**

- [x] Note parser has been updated if structured phrases change

**Code Quality**

- [x] 4-space indents - convert any tabs to spaces
- [x] Use public class field syntax for binding functions - ex. handleChange = () => { ... };
- [x] Code is commented

**Tests**

- [x] Existing tests passed 
- N/A Added Enzyme-UI tests for slim mode 
- N/A Added Enzyme-UI tests for full mode
- N/A Added backend tests for new functionality added
- [x] Added note parser examples to test new functionality

## Reviewer 1: Name

- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

## Reviewer 2: Name

- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

**Comment by nicoleng12** [comment]
Addressed G1. Updated documentation for inserter shortcuts to include the new attributes added in this branch (https://github.com/FluxNotes/flux/wiki/Declarative-Shortcut-Format)

**Comment by gregquinn2001** [comment]
[GQ2] When note is closed, the stop medication message in medication graphical visualization should be on same line with medication name.
[GQ3] Create a JIRA to handle unsigned status. putting stop medication in an unsigned note should have a visual indication that the stop is unsigned like we have for data values in tabular and narrative views

**Comment by nicoleng12** [comment]
@gregquinn2001  Addressed GQ2 and also created a jira to update the medication range chart to make the name property optional: https://jira.mitre.org/browse/ASCODCP-1184. Turned GQ3 into a jira as requested: https://jira.mitre.org/browse/ASCODCP-1185

**Comment by jafeltra** [comment]
I just started reviewing this, but I have a question. Now that we visualize stopping medications by removing the chart, it differs visually from the other medications that are not active. Other medications that are not active don't have an indication that they are not active, but stopped medications do. Should we add something to the visualization for not active medications? Or maybe we need to update the JSON to correctly stop the medications so they use the visualization Nicole just implemented. @gregquinn2001 what do you think? If we want this, I think it would be a separate Jira.

File: src/lib/MedicationInformationService.jsx [hunk 1]
```diff
@@ -1,3 +1,20 @@
+import codeableConceptUtils from '../model/CodeableConceptUtils.jsx';
+
+// List of medications 
+// TODO: Add more medications
```
Comment (jafeltra) [inline src/lib/MedicationInformationService.jsx]: Is this TODO something that will happen in future tasks?

Comment (nicoleng12) [reply src/lib/MedicationInformationService.jsx]: Yes, as it is needed, we can add more medications here.

File: src/shortcuts/ShortcutManager.js [hunk 2]
```diff
@@ -251,6 +251,8 @@ class ShortcutManager {
 
     getValidChildShortcutsInContext(context, recurse = false) {
         const currentContextId = context.getId();
+        // console.log("current context id: " + currentContextId);
```
Comment (jafeltra) [inline src/shortcuts/ShortcutManager.js]: There's a few commented out console.logs in the file. Do we need to keep those?

**Comment by jafeltra** [comment]
[JA1] Is there a way to not have the "stopped on..." line so far from the name of the medication when a note is open? It seems strange to have it so far way from the title. It also looks a little weird since the "stopped on..." line is centered but the name of the medication is not.

[JA2] The following UI tests failed:
1. Clicking "@condition" and choosing "Invasive ductal carcinoma of breast" creates a new condition section in the context tray.
2. Clicking "@condition" and choosing multiple conditions does not allow user to select other conditions besides the current condition from the context tray in the condition section.
3. Clicking the data visualization buttons changes the visualizer used
4. Clicking the data visualization buttons changes the visualizer used

[JA3] Using #staging is broken. It gives me the following error: 
Uncaught TypeError: Cannot read property 'constructor' of undefined
    at FluxCondition.js:163

[JA4] Typing #disease status #imaging does not update the targeted data panel with imaging. Clicking the options in the Context Tray does seem to work.

[JA5] #ER is broken. Typing and clicking both don't work. I have not tested every shortcut, but I'm guessing this is the case for others as well.  

This is more of a general comment, but it seems like typing in the editor has gotten a little bit laggy again. I'm not entirely sure where that came from.

**Comment by jafeltra** [comment]
[JA6] Debra (/demo1) does not have any active medications, meaning there is nothing to open in a portal. It seems a little weird when no portal opens, but I'm not sure what a better way to handle that is. More concerning immediately, if you hit enter after trying to open the empty portal, you get the following error:
Uncaught TypeError: Cannot read property 'context' of undefined
    at Object.FluxNotesEditor._this.onPortalSelection [as onSelected] (FluxNotesEditor.jsx:329)

File: src/shortcuts/ShortcutManager.js (hunk 2 above)
Comment (nicoleng12) [reply src/shortcuts/ShortcutManager.js]: No, good catch. I removed them and committed

File: src/lib/MedicationInformationService.jsx (hunk 1 above)
Comment (nicoleng12) [reply src/lib/MedicationInformationService.jsx]: I added some more comments specifying that this is mainly used by the note parser right now

**Comment by nicoleng12** [comment]
Made a jira for JA1: https://jira.mitre.org/browse/ASCODCP-1184

**Comment by nicoleng12** [comment]
JA3 and JA5 are fixed now. As part of JA5 I also fixed PR and HER2 which was broken for the same reason that ER was broken. JA4 is expected behavior because cannot have rationale without status. This is the same behavior that is on fluxnotes.org. Created a jira for JA6: https://jira.mitre.org/browse/ASCODCP-1190

**Comment by jafeltra** [comment]
This looks good! All the JA's were either fixed or there are Jiras to cover them. JA2 apparently is broken on master too, so we can address those separately.

**Comment by jafeltra** [comment]
JA7: Add a note parser test to test the new shortcut and parsing features you added.

**Comment by nicoleng12** [comment]
Addressed JA7

**Comment by nicoleng12** [comment]
Rebased

**Comment by jafeltra** [comment]
Looks good! Test passed! Nice job!

**Pull request merged**